lib/rubygems.rb

```diff
@@ -590,22 +590,25 @@ def self.load_path_insert_index
 
     index = $LOAD_PATH.index RbConfig::CONFIG['sitelibdir']
 
-    index
+    index || 0
+  end
+
+  ##
+  # The number of paths in the `$LOAD_PATH` from activated gems. Used to
+  # proritize `-I` and `ENV['RUBYLIB`]` entries during `require`.
+
+  def self.activated_gem_paths
+    @activated_gem_paths ||= 0
   end
 
   ##
   # Add a list of paths to the $LOAD_PATH at the proper place.
 
   def self.add_to_load_path(*paths)
-    insert_index = load_path_insert_index
+    @activated_gem_paths += paths.size
 
-    if insert_index
-      # gem directories must come after -I and ENV['RUBYLIB']
-      $LOAD_PATH.insert(insert_index, *paths)
-    else
-      # we are probably testing in core, -I and RUBYLIB don't apply
-      $LOAD_PATH.unshift(*paths)
```

Comment:
Honestly, the `$LOAD_PATH.unshift(*paths)` should be the only logic which should be used here, because this is wrong:

~~~
$ ruby -Ifoo -e 'gem "psych"; puts $LOAD_PATH'
/foo
/builddir/.gem/ruby/2.7.0/gems/psych-3.1.0/lib
/builddir/.gem/ruby/2.7.0/extensions/x86_64-linux/2.7.0-static/psych-3.1.0
/usr/local/lib/ruby/site_ruby/2.7.0
/usr/local/lib/ruby/site_ruby/2.7.0/x86_64-linux
/usr/local/lib/ruby/site_ruby
/usr/local/lib/ruby/vendor_ruby/2.7.0
/usr/local/lib/ruby/vendor_ruby/2.7.0/x86_64-linux
/usr/local/lib/ruby/vendor_ruby
/usr/local/lib/ruby/2.7.0
/usr/local/lib/ruby/2.7.0/x86_64-linux
~~~

the `-Ifoo` was specified earlier, therefore it should have lower priority then the `gem "psych"` call.

Reply:
Also, I am not convinced that the original reason as #1866 is even correct, because default gems are simply loaded from the ruby stdlib and therefore every `-I` option has *always* precedence, so I think the original test case must be flawed.

Reply:
Sorry, my initial example was probably not complete enough because it is not obvious that the activated Psych is user installed. Here is more complete example:

~~~
$ gem list | grep psych
psych (3.1.0)

$ ruby -Ifoo -e 'gem "psych"; puts $LOAD_PATH'
/foo
/builddir/.gem/ruby/2.7.0/gems/psych-3.1.0/lib
/builddir/.gem/ruby/2.7.0/extensions/x86_64-linux/2.7.0-static/psych-3.1.0
/usr/local/lib/ruby/site_ruby/2.7.0
/usr/local/lib/ruby/site_ruby/2.7.0/x86_64-linux
/usr/local/lib/ruby/site_ruby
/usr/local/lib/ruby/vendor_ruby/2.7.0
/usr/local/lib/ruby/vendor_ruby/2.7.0/x86_64-linux
/usr/local/lib/ruby/vendor_ruby
/usr/local/lib/ruby/2.7.0
/usr/local/lib/ruby/2.7.0/x86_64-linux

$ rm -rf ~/.gem/

$ ruby -Ifoo -e 'gem "psych"; puts $LOAD_PATH'
/foo
/usr/local/lib/ruby/site_ruby/2.7.0
/usr/local/lib/ruby/site_ruby/2.7.0/x86_64-linux
/usr/local/lib/ruby/site_ruby
/usr/local/lib/ruby/vendor_ruby/2.7.0
/usr/local/lib/ruby/vendor_ruby/2.7.0/x86_64-linux
/usr/local/lib/ruby/vendor_ruby
/usr/local/lib/ruby/2.7.0
/usr/local/lib/ruby/2.7.0/x86_64-linux
~~~

The first example is the same as the one above with locally installed psych overriding the default gem. Then I delete user the `.gem` directory, which removes the installed gem and go with defaults. As you can see, there are no paths inserted anywhere for the default gem, because there are not paths to be inserted and therefore I believe the whole logic is there just to satisfy the wrong test case not reflecting the reality.

Reply:
Regarding your first comment, that's how rubygems require has been working for years, gems have always been added right before the default `$LOAD_PATH`, not at the front, and the reason has always been that `-I` CLI flags should take precedence. If you feel that historical behavior should be changed, feel free to open an issue to discuss.

> Also, I am not convinced that the original reason as #1866 is even correct, because default gems are simply loaded from the ruby stdlib and therefore every -I option has always precedence, so I think the original test case must be flawed.

Before #1866, if you had `fileutils.rb` file in your current folder and specified `-I.`, the file in your current folder would be ignored because the `fileutils` default gem would get activated and the file in the default `$LOAD_PATH` would get loaded.

Reply:
Sorry, our comments crossed, but I don't see how the last comment you posted is related to this ticket or any of the related tickets. Can you elaborate a bit? What you're doing there (removing a gem directly without `gem uninstall` and thus not removing its gemspec) is something I had never tried and we just learnt how rubygems behaves there :laughing:. But I don't quite see the relation.

Reply:
Sorry, my comment was not accurate. The file in your currrent folder will be loaded, but the default gem with the same name will also be activated, and from the point of view of #1866, thar is unexpected. I understand that it is debatable though and I'm open to discussing switching back to the old behavior which I believe is what you're proposing?

Reply:
> that's how rubygems require has been working for years, gems have always been added right before the default $LOAD_PATH, not at the front, and the reason has always been that -I CLI flags should take precedence.

Ah, you are right. Sorry, I have been mistaken.

My entire point is that if you take high level look at what RubyGems do, the `gem` call is essentially just modification of `$LOAD_PATH`. From that POV call of `gem 'psych'` for default gem is no-op (second example) while it modifies the `$LOAD_PATH` for non default gem (first example). The #1866 talks just about "default" gems. If the gem activation for default gem is no-op, then the `-I` option has to automatically take the precedence.

I understand the situation is a bit more complex behind curtain, but maybe it is overly complex. Considering that the test case just tries to emulate what is the real situation, I have just low confidence that the test case is correct. But I have not done any deep investigation.

> (removing a gem directly without gem uninstall and thus not removing its gemspec)

Here I remove the whole directory from user home, therefore I remove everything, the gem, the gemspec, the documentation, the gem cache, etc, making sure I reset my environment to pristine conditions. I could do `gem uninstall` of course, but that would still left some leftovers behind.

Reply:
Looking close into #1868, I think that call to `gem_original_require(resolved_path)` means that the `Kernel#require` is called with completely different path then originally requested, so `require "psych"` might RubyGems transform to `require "some/path/psych"` and that is not acceptable IMO.

Reply:
Sorry, I missed that your example was removing the whole gem directory. Then I'm not sure why activating psych is not failing, I would've expected it to fail? :thinking:. Anyways, I feel it's only tangentially related to this issue?

 > My entire point is that if you take high level look at what RubyGems do, the gem call is essentially just modification of $LOAD_PATH. From that POV call of gem 'psych' for default gem is no-op (second example) while it modifies the $LOAD_PATH for non default gem (first example). The #1866 talks just about "default" gems. If the gem activation for default gem is no-op, then the -I option has to automatically take the precedence.

> I understand the situation is a bit more complex behind curtain, but maybe it is overly complex. Considering that the test case just tries to emulate what is the real situation, I have just low confidence that the test case is correct. But I have not done any deep investigation.

Yes, it is more complex. And yes, maybe it's overly complex. Gem activation for default gems does not add paths to the `$LOAD_PATH` but it's definitely not a no-op. As I said, I'm happy to try simplify/change things so that they are simpler. But again, separate discussion. Here I'm just fixing two bugs, and one of those were reported by you, so I would guess you are ok with fixing it?

> Looking close into #1868, I think that call to gem_original_require(resolved_path) means that the Kernel#require is called with completely different path then originally requested, so require "psych" might RubyGems transform to require "some/path/psych" and that is not acceptable IMO.

This is what ruby's C `require` does internally when you pass a string to it, instead of a path.I wouldn't think that a reason per se for not being acceptable, but I'm happy to discuss it. Again, I feel it fits a separate discussion better.

Can you give me your opinion om whether you would accept, reject, or propose any specific modifications to this change set?

Reply:
I'll hold this PR until I have a look at your #3647 report. Thanks, neat report! :)

Reply:
With the case presented by @voxik in #3647 it becomes more clear that he's right about

> Looking close into #1868, I think that call to gem_original_require(resolved_path) means that the Kernel#require is called with completely different path then originally requested, so require "psych" might RubyGems transform to require "some/path/psych" and that is not acceptable IMO.

not being acceptable. Because with that code, the `$LOADED_FEATURES` cache of standard require is no longer hit. And there's no way around that unless we reimplement that kind of caching ourselves which seems like... Not the way to go.

I'm trying to revert #1868 in #3648 as an alternative.

Reply:
Ok, so I had an idea and ended up adding a couple of extra commits here to also fix #3647. Now there's no manual `$LOAD_PATH` traversal whatsoever for every require, so this should definitely fix #3631, and also there's no more `gem_original_require` calls with a "modified path".

Reply:
Concerns should've been addressed now.